### PR TITLE
[prover] spec for `big_vector`

### PR DESCRIPTION
**Pull request by chen-robert**
### Description

Here we add specifications for `big_vector`. In particular, we specify
1. Critical data invariants about the structure of `big_vector`
2. Well-scoped `aborts_if` clauses 

The goal is for this to serve as a model for further data structure specifications. 

As an aside, we also note potentially confusing behavior with `table::spec_get`. It might make sense to add a `safe_spec_get` variant that checks for the existence of the key instead of returning garbage. 

### Test Plan

Use the Aptos prover.

### CODE REVIEW

**Comment by junkil-park** [comment]
@chen-robert , to pass `Lint+Test`, please regenerate the documentation by running `cargo test --release -p aptos-framework`.

**Review by clay-aptos — APPROVED** [review]

**Comment by junkil-park** [comment]
And, this is the error message from the Prover test: https://github.com/aptos-labs/aptos-core/actions/runs/4557715742/jobs/8039715206?pr=7209#step:8:378

**Review by junkil-park — APPROVED** [review]

**Comment by junkil-park** [comment]
>As an aside, we also note potentially confusing behavior with table::spec_get. It might make sense to add a safe_spec_get variant that checks for the existence of the key instead of returning garbage.

Actually, this is intended. `spec_get` is supposed to be a partial function, which returns a garbage (undefined) value when the key is absent in the table. Likewise, `global` is also a partial function, and `global<R>(a)` is undefined when `R` does not exist under `a`. So, one may need to use `exists<R>(a)` if they are not sure if it actually exists.

It has been the user's responsibility for the implication of the partial semantics of those functions. One missing but desired feature of Prover is to warn users if they are using partial functions that can potentially return undefined values.

**Comment by junkil-park** [comment]
After this PR lands, I will port some more spec of `big_vector` from my outstanding PR https://github.com/aptos-labs/aptos-core/pull/6746.

**Review by rahxephon89 — APPROVED** [review]

**Comment by 1nc0gn170** [comment]
@junkil-park When can we expect this PR to be merged?

**Comment by sh1v-sh4nk** [comment]
@davidiw @clay-aptos @wrwg @movekevin @saharct Please approve these changes.

**Comment by junkil-park** [comment]
It doesn't get merged due to the lint test failure: https://github.com/aptos-labs/aptos-core/actions/runs/4630114301/jobs/8299093657?pr=7209#step:5:79

Could you fix the whitespace issue?

File: aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move [hunk 1]
```diff
@@ -131,14 +136,16 @@ module aptos_std::big_vector {
...
+            assert cur_bucket_index == num_buckets;
+        };
 
         // Shrink the table if the last vector is empty.
         let last_bucket = table_with_length::borrow_mut(&mut v.buckets, num_buckets - 1);
         if (vector::is_empty(last_bucket)) {
             move last_bucket;
             vector::destroy_empty(table_with_length::remove(&mut v.buckets, num_buckets - 1));
         };
-        v.end_index = v.end_index - 1;
```
Comment (movekevin) [inline aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move:141]: Why did this get deleted?

Comment (chen-robert) [reply aptos-move/framework/aptos-stdlib/sources/data_structures/big_vector.move:141]: This was moved above to line 123 to make the invariant more straightforward

**Pull request merged**